src/outputs/restart.cpp

```diff
@@ -469,6 +469,111 @@ void RestartOutput::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm)
+    local_count[3] -= 1;
+    global_count[3] -= 1;
+
+    local_count[2] += 1;
+    global_count[2] += 1;
+    auto vLocalSpaceY = H5Screate_simple(5, local_count, NULL);
+    auto vGlobalSpaceY = H5Screate_simple(5, global_count, NULL);
+    local_count[2] -= 1;
+    global_count[2] -= 1;
+
+    local_count[1] += 1;
+    global_count[1] += 1;
+    auto vLocalSpaceZ = H5Screate_simple(5, local_count, NULL);
+    auto vGlobalSpaceZ = H5Screate_simple(5, global_count, NULL);
+    local_count[1] -= 1;
+    global_count[1] -= 1;
+
+    hsize_t offset = (nx1 + 1) * (nx2 + 1) * (nx3 + 1);
+    hsize_t index1 = 0;
+    hsize_t index2 = offset;
+    hsize_t index3 = 2 * offset;
+    for (auto &pmb : pm->block_list) { // for every block1
+      auto ci =
+          MeshBlockDataIterator<Real>(pmb->meshblock_data.Get(), output_params.variables);
+      for (auto &v : ci.varsFace) {
+        std::string name = v->label();
+        if (name.compare(vWriteName) == 0) {
+          // copy data to host
+
+          // Load x direction
+          auto v_x = v->Get(1).GetHostMirrorAndCopy();
+          LOADVARIABLEONE(index1, dataF, v_x, out_ib.s, out_ib.e + 1, out_jb.s, out_jb.e,
+                          out_kb.s, out_kb.e, vlen);
+          if (pm->ndim > 1) {
+            // Load y direction
+            auto v_y = v->Get(2).GetHostMirrorAndCopy();
+            LOADVARIABLEONE(index2, dataF, v_y, out_ib.s, out_ib.e, out_jb.s,
+                            out_jb.e + 1, out_kb.s, out_kb.e, vlen);
+          }
+          if (pm->ndim > 2) {
+            // Load z direction
+            auto v_z = v->Get(3).GetHostMirrorAndCopy();
+            LOADVARIABLEONE(index3, dataF, v_z, out_ib.s, out_ib.e, out_jb.s, out_jb.e,
+                            out_kb.s, out_kb.e + 1, vlen);
+          }
+          break;
+        }
+      }
+    }
+    // write datasets to file
+    // X direction
+    {
+      std::string nameX = vWriteName + "_x1";
+      local_count[3] += 1;
+      WRITEH5SLAB2(nameX.c_str(), dataF, file, local_start, local_count, vLocalSpaceX,
+                   vGlobalSpaceX, property_list);
+      local_count[3] -= 1;
+    }
+    if (pm->ndim > 1) {
+      dataF += offset;
+      std::string nameX = vWriteName + "_x2";
+      local_count[2] += 1;
+      WRITEH5SLAB2(nameX.c_str(), dataF, file, local_start, local_count, vLocalSpaceY,
+                   vGlobalSpaceY, property_list);
+      local_count[2] -= 1;
+    }
+    if (pm->ndim > 2) {
+      dataF += offset;
+      std::string nameX = vWriteName + "_x3";
+      local_count[1] += 1;
+      WRITEH5SLAB2(nameX.c_str(), dataF, file, local_start, local_count, vLocalSpaceZ,
+                   vGlobalSpaceZ, property_list);
+      local_count[1] -= 1;
+    }
+    // close data spaces
+    H5Sclose(vLocalSpaceX);
+    H5Sclose(vGlobalSpaceX);
+    H5Sclose(vLocalSpaceY);
+    H5Sclose(vGlobalSpaceY);
+    H5Sclose(vLocalSpaceZ);
+    H5Sclose(vGlobalSpaceZ);
+  }
+
```

Comment:
This is almost line-for-line duplicated between restart and output. Can it be abstracted a bit. Even if it has to eventually change, it'd be better to change things in only one spot.